assets/app/scripts/services/data.js

```diff
@@ -199,15 +125,269 @@ angular.module('openshiftConsole')
+//                    list.by('metadata.uid'))
+//            event:  specific event that caused this call ("ADDED", "MODIFIED",
+//                    "DELETED", or null) callbacks can optionally use this to 
+//                    more efficiently process updates
+//            obj:    specific object that caused this call (may be null if the
+//                    entire list was updated) callbacks can optionally use this 
+//                    to more efficiently process updates
+// opts:      options 
+//            poll:   true | false - whether to poll the server instead of opening
+//                    a websocket. Default is false.
+//            pollInterval: in milliseconds, how long to wait between polling the server
+//                    only applies if poll=true.  Default is 5000.
+//
+// returns handle to the watch, needed to unwatch e.g.
+//        var handle = DataService.watch(type,context,callback[,opts])
+//        DataService.unwatch(handle)
+  DataService.prototype.watch = function(type, context, callback, opts) {
+    this._watchCallbacks(type, context).add(callback);
+
+    if (this._watch(type, context) && this._resourceVersion(type, context)) {
+      callback(this._data(type, context));
+    }
+    else if (this._list(type, context)) {
+      // no-op, our callback will get called when listOperation completes
+    }
+    else {
+      this._startListOp(type, context, opts);
+    }
+
+    // returned handle needs type, context, and callback in order to unwatch
+    return {
+      type: type,
+      context: context,
+      callback: callback,
+      opts: opts
+    };
+  };
+
+  DataService.prototype.unwatch = function(handle) {
+    var type = handle.type;
+    var context = handle.context;
+    var callback = handle.callback;
+    var opts = handle.opts;
+    var callbacks = this._watchCallbacks(type, context);
+    callbacks.remove(callback);
+    if (!callbacks.has()) {
+      if (opts && opts.poll) {
+        clearTimeout(this._watchPollTimeouts(type, context));
+        this._watchPollTimeouts(type, context, null);
+      }
+      else {
+        this._watchWebsockets(type, context).close();
+        this._watchWebsockets(type, context, null);
+      }
+      
+      this._watch(type, context, false);
+    }
+  };
+
+
+  DataService.prototype._watchCallbacks = function(type, context) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!this._watchCallbacksMap[key]) {
+      this._watchCallbacksMap[key] = $.Callbacks();
+    }
+    return this._watchCallbacksMap[key];
+  };
+
+  DataService.prototype._listCallbacks = function(type, context) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    if (!this._listCallbacksMap[key]) {
+      this._listCallbacksMap[key] = $.Callbacks();
+    }
+    return this._listCallbacksMap[key];
+  };
+
+  // maybe change these
+  DataService.prototype._watch = function(type, context, op) {
```

Comment:
rename to _watchInFlight, _listInFlight? This was confusing... I thought _list was actually starting the operation